flink-runtime/src/main/java/org/apache/flink/runtime/util/KeyedBudgetManager.java

```diff
@@ -0,0 +1,238 @@
+ */
+@ThreadSafe
+public class KeyedBudgetManager<K> {
+	private final Map<K, Long> maxBudgetByKey;
+
+	private final long defaultPageSize;
+
+	private final long totalNumberOfPages;
+
+	@GuardedBy("lock")
+	private final Map<K, Long> availableBudgetByKey;
+
+	private final Object lock = new Object();
+
+	public KeyedBudgetManager(Map<K, Long> maxBudgetByKey, long defaultPageSize) {
+		Preconditions.checkNotNull(maxBudgetByKey);
+		Preconditions.checkArgument(defaultPageSize > 0L, "The default page size has to be greater than zero");
+
+		this.maxBudgetByKey = new HashMap<>(maxBudgetByKey);
+		this.availableBudgetByKey = new HashMap<>(maxBudgetByKey);
+		this.defaultPageSize = defaultPageSize;
+		this.totalNumberOfPages = calculateTotalNumberOfPages(maxBudgetByKey, defaultPageSize);
+	}
+
+	public long getDefaultPageSize() {
+		return defaultPageSize;
+	}
+
+	/**
+	 * Tries to acquire budget for a given key.
+	 *
+	 * <p>No budget is acquired if it was not possible to fully acquire the requested budget.
+	 *
+	 * @param key the key to acquire budget from
+	 * @param size the size of budget to acquire from the given key
+	 * @return the fully acquired budget for the key or max possible budget to acquire
+	 * if it was not possible to acquire the requested budget.
+	 */
+	public long acquireBudgetForKey(K key, long size) {
+		Preconditions.checkNotNull(key);
+		Either<Map<K, Long>, Long> result = acquirePagedBudgetForKeys(Collections.singletonList(key), size, 1L);
+		return result.isLeft() ? result.left().get(key) : result.right();
+	}
+
+	/**
+	 * Tries to acquire budget for given keys which equals to the number of pages times default page size.
+	 *
+	 * <p>See also {@link #acquirePagedBudgetForKeys(Iterable, long, long)}
+	 */
+	public Either<Map<K, Long>, Long> acquirePagedBudget(Iterable<K> keys, long numberOfPages) {
+		return acquirePagedBudgetForKeys(keys, numberOfPages, defaultPageSize);
+	}
+
+	/**
+	 * Tries to acquire budget which equals to the number of pages times page size.
+	 *
+	 * <p>The budget will be acquired only from the given keys. Only integer number of pages will be acquired from each key.
+	 * If the next page does not fit into the available budget of some key, it will try to be acquired from another key.
+	 * The acquisition is successful if the acquired number of pages for each key sums up to the requested number of pages.
+	 * The function does not make any preference about which keys from the given keys to acquire from.
+	 *
+	 * @param keys the keys to acquire budget from
+	 * @param numberOfPages the total number of pages to acquire from the given keys
+	 * @param pageSize the size of budget to acquire per page
+	 * @return the acquired number of pages for each key if the acquisition is successful (either left) or
+	 * the total number of pages which were available for the given keys (either right).
+	 */
+	Either<Map<K, Long>, Long> acquirePagedBudgetForKeys(Iterable<K> keys, long numberOfPages, long pageSize) {
```

Comment:
Yes `BudgetAllocationException` would be more accurate, and my suggestion comes from the fact that now `MemoryManager.reserveMemory` also throws a `MemoryAllocationException`, and I agree change the method signature could better reflect the case.

Personally I think the semantic is explicit to return the acquired budget when succeed and throw exception when fail. Having an internal functional `Try` allows incorrect handling in upper caller, like if `MemoryManager.reserveMemory` doesn't check the returning value and continues, no warning/error will happen but there would be memory leak. Throwing an explicit exception will also make the test coverage tool working better (thus help easier maintenance), for the existing implementation I need to manually verify whether `Either.right` is checked in test case since it won't be reflected by test coverage.

Reply:
Ok, I added `AllocationResult` which throws as exception if result is not checked but the wrong is queried.